command/debug.go

```diff
@@ -0,0 +1,683 @@
+	// Version contains cluster meta information
+	dir := "version"
+	err := c.mkdir(dir)
+	if err != nil {
+		return err
+	}
+
+	self, err := client.Agent().Self()
+	if err != nil {
+		return fmt.Errorf("error agent self: %s", err.Error())
+	}
+	c.writeJSON(dir, "agent-self.json", self)
+
+	// Fetch data directly from consul and vault. Ignore errors
+	var consul, vault string
+
+	m, ok := self.Config["Consul"].(map[string]interface{})
+	if ok {
+		raw := m["Addr"]
+		consul, _ = raw.(string)
+		raw = m["EnableSSL"]
+		ssl, _ := raw.(bool)
+		if ssl {
+			consul = "https://" + consul
+		} else {
+			consul = "http://" + consul
+		}
+	}
+
+	m, ok = self.Config["Vault"].(map[string]interface{})
+	if ok {
+		raw := m["Addr"]
+		vault, _ = raw.(string)
+	}
+
+	c.collectConsul(dir, consul)
+	c.collectVault(dir, vault)
+
+	c.startMonitors(client)
+	c.collectPeriodic(client)
+	c.collectPprofs(client)
+
+	return nil
+}
+
+// path returns platform specific paths in the tmp root directory
+func (c *DebugCommand) path(paths ...string) string {
+	ps := []string{c.collectDir}
+	ps = append(ps, paths...)
+	return filepath.Join(ps...)
+}
+
+// mkdir creates directories in the tmp root directory
+func (c *DebugCommand) mkdir(paths ...string) error {
+	return os.MkdirAll(c.path(paths...), 0755)
+}
+
+// startMonitors starts go routines for each node and client
+func (c *DebugCommand) startMonitors(client *api.Client) {
+	for _, id := range c.nodeIDs {
+		go c.startMonitor("client", "node_id", id, client)
+	}
+
+	for _, id := range c.serverIDs {
+		go c.startMonitor("server", "server_id", id, client)
+	}
+}
+
+// startMonitor starts one monitor api request, writing to a file. It blocks and should be
+// called in a go routine. Errors are ignored, we want to build the archive even if a node
+// is unavailable
+func (c *DebugCommand) startMonitor(path, idKey, nodeID string, client *api.Client) {
+	c.mkdir(path, nodeID)
+	fh, err := os.Create(c.path(path, nodeID, "monitor.log"))
+	if err != nil {
+		return
+	}
+	defer fh.Close()
+
+	qo := api.QueryOptions{
+		Params: map[string]string{
+			idKey:       nodeID,
+			"log_level": c.logLevel,
+		},
+	}
+	outCh, errCh := client.Agent().Monitor(c.stopCh, &qo)
+	for {
+		select {
+		case out := <-outCh:
+			if out == nil {
+				continue
+			}
+			fh.Write(out.Data)
+			fh.WriteString("\n")
+
+		case err := <-errCh:
+			fh.WriteString(fmt.Sprintf("Error monitoring: %s\n", err.Error()))
+			return
+
+		case <-c.stopCh:
+			return
+		}
+	}
+}
+
+// collectPprofs captures the /agent/pprof for each listed node
+func (c *DebugCommand) collectPprofs(client *api.Client) {
+	for _, n := range c.nodeIDs {
+		c.collectPprof("client", n, client)
+	}
+
+	for _, n := range c.serverIDs {
+		c.collectPprof("server", n, client)
+	}
+
+}
+
+// collectPprof captures pprof data for the node
+func (c *DebugCommand) collectPprof(path, id string, client *api.Client) {
+	opts := api.PprofOptions{Seconds: 1}
```

Comment:
This is at least a little silly, it would be more useful to both configure this value and maybe do profiles in intervals like we do capturing state. There is a concern that only one profile can be run on an agent at a time, which causes some problems. I think this will still capture useful data, and I'm planning on improving this in a followup.